### DISCUSSION TITLE
关于milvus1.1.1版本下分布式部署模式基于mishards的服务路由策略设计问题

### DISCUSSION THREAD

**Discussion by tinygiants-tech** [milvus-io/milvus#7101] [Q&A and General discussion]
zilliz的工程师你们好：
       最近在思考mishards集群中间件的个性化路由策略，考虑到实际应用场景下可能存在存量数据存储节点爆满后扩容的方案设计，可能仍需保证和存量的数据是相同的collection和partition，对于增量插入请求的数据自动装载内存到新节点，不知道是否能够通过个性化路由插件设计实现该功能，另外，考虑一致性哈希在扩容是会对存量数据产生影响，涉及单个segment段数据的迁移，公司这边打算直接通过个性化路由插件设计来实现扩容，但希望在扩容的时候对存量数据不做迁移处理，保证存量数据节点不动，新数据自动被装载到新节点，期待你们的回复，谢谢！

**Comment by yhmo** [comment]
要改变mishards数据分发逻辑，需要仿照 /milvus/shards/mishards/router/plugins/file_based_hash_ring_router.py 这个文件写一个类似的。

在这个文件的 _route() 方法中：
- 从47行到113行是从MySQL里查询有多少segments需要被加载
- 从115行到140行是将每个segment的id作为字符串hash得到它应该加载到哪个read-only 节点
- 第117行的servers = self.readonly_topo.group_names是拿到read-only节点列表，这里的readonly_topo也是一个插件，它可能是/milvus/shards/discovery/plugins/kubernetes_provider.py， 也可能是/milvus/shards/discovery/plugins/static_provider.py。在KubernetesProvider这个类中，你可以看到它会根据Kubernetes事件来增加或者减少read-only节点。
- 第125行的ring.get_node()是用hash ring来决定segment所属的read-only节点

对于你们的需求来说，我觉的要做如下修改：
- 另写一个负责分发segment的类，替代hash_ring.py，这个类要知道哪个节点是新节点，要知道哪些节点已经爆满
- 仿照file_based_hash_ring_router重写一个router，将从115行到140行这段代码换掉
